Iteration protocol change

Keno opened the pull request:
I imported @vtjnash's lowering change and then went ahead and rewrote the base iterator wrappers to use the new iteration protocol instead so people can see it in action. This passes the iterators test, but I haven't run other tests yet, so I suspect some fail.

The strategy here basically what is described in #18823, with the exception of using unions rather than nullables. Additionally, I kept `done` around as an optional addition (the semantic guarantee
is that if done is not an error, then it determines whether or not the next call to iterate will return true). This was necessary, because the `isempty` function was derived from the iteration protocol previously, which is no longer possible with the new code. However, after playing with this, it is not actually clear to me that any use of `done` other than to implement `isempty` is useful with the new iteration protocol, so maybe the correct thing here is just to have iterators optionally implement `isempty` just as they would `length` now (for iterators where this is determinable).

Couple of things on my todo list:
- [x] Fix other tests
- [x] Write documentation for new iteration protocol
- [x] Add lookahead iterator that adds a reliable `done` method to any iterator by prefetching an element and keeping it in the state (essentially the generic version of the trick people were using to implement `done` for such iterators before)

There was also some performance concerns needed to be looked at (and @JeffBezanson and @vtjnash wanted to help out with), so hopefully this branch can help with that.

Comments and feedback greatly appreciated (again though, WIP, not supposed to pass tests, etc, so there's probably bugs, but hopefully this is useful to get a feel for what this change does).

Keno committed:
Change iteration protocol

This changes the iteration protocol from `start`/`next`/`done` to `iterate`.
The new lowering of a for loop is as follows:

```
for x in itr
    ...
end
```

becomes

```
next = iterate(itr)
while next !== nothing
    x, state = next::Tuple{Any, Any}
    ...
    next = iterate(itr, state)
end
```

The semantics are as apparent from the above lowering. `iterate` returns
either `nothing` or a tuple of value and state. The state is passed
to any subsequent operation. The first iteration is indicated, by not passing
the second, state argument to the `iterate` method.

Adaptors in both directions are provided to keep the legacy iteration
protocol working for now. However, performance of the legacy iteration
protocol will be severely pessimized.

As an optional add-on for mutable iterators, a new `isdone` function is
provided. This function is intended as an O(1) approximate query for
iterator completion, where such a calculation is possible without mutation
and/or is significantly faster than attempting to obtain the element itself.
The function makes use of 3-value logic. `missing` is always an acceptable
answer, in which case the caller should go ahead and attempt the iteration
to obtain a definite result. If the result is not `missing`, it must be
exact (i.e. if true, the next call to iterate must return `nothing`, if
false it must not return nothing).

Keno commented:
Another thing to discuss is whether the absence of `start` in this proposal is an advantage or a disadvantage. In essence, this proposal uses the absence of a `state` argument to `iterate` as way to write the separate method based on the observation that for simple iterators, the initial state argument is often trivial:
```
iterate(x::Array, i=1) = i > length(x) ? nothing : (a[i], i+1)
```
However, it has the distinct disadvantage that it can't carry any state. I didn't appreciate this aspect before (since it's sort of implicit in our current iteration protocol), but there are certain iterators that want to do some pre-processing at the beginning of an iteration, e.g. imagine:
```
struct EachLineFile; path::String; end
start(f::File) = (itr = EachLine(open(f.path)); (itr, start(itr)))
iterate(f::File, state=start(f)) = iterate(state...)
```
which you can of course do just as above for the first argument, but I think it might validly be a problem for separation of concerns. More practical considerations might be something like #22467 (which may also be the case for the EachLineFile operator above depending how you implement it).

On the other hand, there may be iterators for which constructing a valid, type-stable `state` object is not possible without obtaining the element.

I think the best solution I can currently imagine is making the iteration protocol something like the following:
```
y = iterate(coll, start(coll))
while y !== nothing
    val, state = y
    # body
    y = iterate(coll, state)
end
```
this has the advantage that it's file if the return type of `start` is different from that of the regular `state` object, because they never get assigned to the same variable.

We could even do something like:
```
start(x::Any) = start
iterate(x::Any, ::typeof(start)) = iterate(x)
```
to preserve the nice one liner above for simple iterators, but I worry about whether that'll be too complicated.

Keno commented:
Another possibility:
```
it = prepare(coll)
y = iterate(it)
while y !== nothing
    val, state = y
    # body
    y = iterate(it, state)
end
```
with
```
prepare(x::Any) = x
```

nalimilan commented:
Leaving breadcrumbs to https://github.com/JuliaLang/julia/pull/23642/commits/ec91f63b18045956db907c0c3c8dbfab6de286ea from PR https://github.com/JuliaLang/julia/pull/23642, where I had to introduce a `MaybeValue` hack to `ProductIterator` to avoid the performance impact of returning a value equal to `nothing` when there's no value to return. It would be nice to be able to get rid of it.

Probably related: @davidanthoff discussed the case of iterators which need to allocate a resource on the first iteration, see https://github.com/JuliaLang/julia/issues/22466.

davidanthoff commented:
I think the ``prepare`` idea might allow me to get rid of the whole ``getiterator`` mess in https://github.com/davidanthoff/IteratorInterfaceExtensions.jl for Query.jl and IterableTables.jl. I'm not sure, need to think about it a bit more, but right now I'm pretty positive. If that was the case, it would be a major, major benefit and simplification for that ecosystem. If we also could solve the resource management issue, boy, would that be wonderful.

I hope this PR will stay open for a while? With the holidays it might be tricky to find the time needed to think about this... (I plan to eat, eat and eat the next couple of days, and do nothing else).

timholy commented:
I had a brief look and it seemed pretty reasonable to me. It might be interesting to try to implement the `CartesianIndices` iteration "natively" (using `iterate`) rather than relying on its `start`/`next`/`done` methods. Given my interest in seeing #9080 put to bed, I can take a crack at that if you don't have time.

With regards to `prepare`, your strategy seems quite reasonable. I'd slightly favor the name `iterator(x)` over `prepare(x)`, as with the latter it's not entirely clear what it's being prepared for. For example, if `x` is a file handle, I could imagine several different kinds of preparation depending on what you're planning to do with that file.

Thanks so much for doing this!

nalimilan commented:
I agree with @timholy it would make sense to use `iterator` instead of `prepare`. That's also really close to `getiterator` from IteratorInterfaceExtensions, which was indeed general enough to possibly warrant inclusion in Base when discussing the registration of the package (https://github.com/JuliaLang/METADATA.jl/pull/11692). The poster child for this is calling `iterator` on a data frame to get an iterator encoding information on the column types even though the data frame itself doesn't. In that case, we don't really "prepare" anything, we just return an iterator which will ensure type stability of the iterations.

Keno commented:
I have more changes locally, and I'm working on finishing up the PR without either of the proposed extensions (since they're extensions, it's fairly easy to do either in addition on top). However, having spent some more time thinking about this, I'm not sure I'm convinced that it is useful or necessary. It seems always possible to encode whatever information into the state object. I guess the fundamental question is what can you do with the state other than pass it back to iterate. For some iterators, we generally consider the ability to copy the state and start iterating at the same point again. However, whether that works for a given iterator is somewhat ill specified at this point. I'll keep thinking about this and finish up the PR in the mean time, but it's not clear to me.

timholy commented:
I agree that it seems like it should be able to bundle everything into `start`. It's been ages since I've read the issue about "iterators that don't know they're done until they try," but it seems like it should be possible to try for the first object and stash it in `state`---the same union-splitting that makes this PR possible should also make that possible.

Keno commented:
On slack, @vtjnash and I came up with the following proposed lowering:
```
it = iterator(coll)
advance = iterate(it...)
while advance !== nothing
   f(head(advance))
   advance = iterate(tail(advance)...)
end
```
@vtjnash also proposes the following extension to handle resource cleanup:
```
it = iterator(x)
hascleanup = isapplicable(cleanup, it)
if hascleanup
    let isclean = Atomic(false)
        finalizer( it -> (atomic_set!(isclean, true) || cleanup(it)), it)
    end
end
while true
...
end
hascleanup && cleanup(it...) # auto-inserted for any break or return
```
There's a couple of things to like about this. With a default implementation of
```
iterator(x::Any) = (x,)
```
this reduces to the original proposal in this PR for simple iterators. An implementation like,
```
iterator(x::MyIt) = (x, 1)
```
could be used to give initial state if necessary (or that could still be put into the iterate function itself).
For an iterator that's self-mutating, we could just have
```
function iterate(x::SelfMutatingIterator)
    mutate!(x)
    eof(x) && return nothing
    x.val, (x,)
end
```
which I think nicely expresses the notion that such an iterator doesn't really have state.

Keno commented:
Slight revision from triage:
```
it = iterator(coll)
advance = iterate(it...)
while advance !== nothing
   f(head(advance))
   advance = iterate(first(it), tail(advance)...)
end
```

vtjnash commented:
That seems to have a typo – what is `first(it)`? (probably my fault for writing the original wrong) Here's the fixed version:
```
it = iterate(iterator(coll)...) # NOTE: this perhaps should simplify to just `iterate(coll)`
while it !== nothing
   do(head(it))
   it = iterate(tail(it)...)
end
```

Keno commented:
No, it's not typo. It's the first element of the tuple returned by iterator.

JeffBezanson commented:
I notice a small readability issue: previously we could write `value, state = next(itr, state)`, but that now becomes `x = iterate(itr, x[2]); x[1], x[2]`. Maybe we could return a named tuple, so that this can be written with `x.value, x.state`?

JeffBezanson commented on base/iterators.jl:
````diff
@@ -435,9 +424,9 @@ julia> collect(Iterators.rest([1,2,3,4], 2))
 ```
 """
 rest(itr,state) = Rest(itr,state)
+rest(itr) = itr
````
This is a bit confusing since often the functions `first` and `rest` are synonyms for `head` and `tail`, i.e. `rest(x)` drops the first element.

Keno replied:
I'm happy to drop this method, though it is useful (see the use in _totuple).

StefanKarpinski commented:
> Maybe we could return a named tuple, so that this can be written with `x.value`, `x.state`?

Why not use a real struct type then? `IterationPair` or `IterationValueState` or something.

vtjnash commented:
The ergonomics of not using a `Tuple` are pretty terrible for the Method signature. I don't want to write:
```julia
(iterate(x, state)::IterationPair) = (x[state], state + 1)
```
Have multiple ways of accessing the return value (`x[1]`, `first(x)`, `x.value`) also likely means test coverage will be worse.

vtjnash commented on src/julia-syntax.scm:
```diff
@@ -2464,40 +2469,33 @@
+            `(block
+                ;; *** either this or force all for loop vars local
+                ,.(map (lambda (r) `(local ,r))
+                  (lhs-vars (cadr (car ranges))))
```
indent

vtjnash commented:
If we want to recover / preserve the existing structure, we could provide something like:
```julia
val, state = @iterate (x, state) || return
```

But this is just an example of normal `Union{T, Nothing}` handling, so it'll benefit from whatever special syntax we eventually provide for that:
```julia
val, state = @nullable iterate(x, state) || return
val, state = iterate?(x, state) ||? return
etc., TBD, ...
```

vtjnash commented on src/julia-syntax.scm:
```diff
@@ -2269,7 +2274,7 @@
                             (list (cadr e))))
                 (first  #t))
-       (expand-for (if first 'while 'inner-while)
+       (expand-for first
```
drop lowering code for `inner-while` also (a few lines above here)

vtjnash commented on base/inference.jl:
```diff
@@ -2268,28 +2268,36 @@ function precise_container_type(@nospecialize(arg), @nospecialize(typ), vtypes::
+strip_iter_union(X::Type{Union{Nothing, T}}) where {T} = T
+strip_iter_union(::Type{Nothing}) = Nothing
+strip_iter_union(::Type{Union{}}) = Union{}
+strip_iter_union(T) = T
```
this'll be very slow (because of compilation), and possibly unreliable (because `T = Union{Nothing, T}` is also a valid solution), use `typesubtract` instead

vtjnash commented on stdlib/Distributed/src/pmap.jl:
```diff
@@ -256,13 +256,18 @@ julia> collect(c)
+    head[i] = y[1]
+    while i < n
+        y = iterate(c, y[2])
+        y == nothing && return (resize!(head, i), ())
```
are you making this intentionally type-unstable?

Keno replied:
Frankly I just wanted to get the tests to pass so I could start looking at the performance, though I don't think it matters too much since it's only used in pmap, which is already fairly complex.

nalimilan commented on base/abstractarray.jl:
```diff
@@ -1979,6 +1985,10 @@ function hash(a::AbstractArray{T}, h::UInt) where T
     if isa(a, AbstractVector) && applicable(-, x2, x1)
         n = 1
         local step, laststep, laststate
```
`laststate` isn't used anymore.

Keno commented:
After much triaging, it seems like the favored approach is to do this changed as originally proposed in this PR. As of yet unresolved is whether to add a canonical `iterator` hook, though given the absence of splatting in this iteration protocol, that wouldn't be allowed to specify an initial state.

StefanKarpinski commented:
Can you write down what the iteration protocol ended up being here?

davidanthoff commented:
> Can you write down what the iteration protocol ended up being here?

That would be really great. I just tried to understand what is being proposed now and can't really figure it out because there are so many proposals in the comments here. Just the kind of simple code example that shows how an iterator consumer would look now would be great.

Keno commented:
The iteration protocol in this PR is:
```
y = iterate(coll)
while y !== nothing
    val, state = y
    # do things with val
    y = iterate(coll, state)
end
```
with some special handling for `continue` to still execute the `iterate`.

davidanthoff commented:
> As of yet unresolved is whether to add a canonical ``iterator`` hook, though given the absence of splatting in this iteration protocol, that wouldn't be allowed to specify an initial state.

I don't understand this comment. Do you mean something like:
````julia
it = some_cool_name(coll) # Could be iterator(coll), but I don't really care :)
y = iterate(it)
while y !== nothing
    val, state = y
    # do things with val
    y = iterate(it, state)
end
````
Something like that would be really great.

I think leaving the resource management aspect out for now is a good call, as far as I can tell that could be added without breaking anything in a julia 1.x, plus that should really be designed in a more comprehensive way, not just for iteration.

Keno commented:
> I don't understand this comment. Do you mean something like:

Yes

StefanKarpinski commented:
Triage is in favor but skeptical about the `iterator` part, noting that we have a much bigger, more general "delegation problem" in the language and that putting a bandaid on this specific delegation issue is worse than just coming up with a good, general solution. @JeffBezanson also objects that this complicates every single generic user of the iteration protocol, and muddies issues about traits and such where it becomes unclear whether one should call the trait functions on the original collection or on the result of calling `iterator` on that collection.

Keno commented:
@StefanKarpinski summarizes the thinking fairly well. There seems to be a large design space for which it is too late to get into 1.0 at the moment. Also some version of this could probably be done non-breakingly (as long as base doesn't start using the hook), so we're punting on that for now.

davidanthoff commented:
Thinking about this a bit more, I completely agree with that assessment.

Keno commented:
Update: I believe I've got all the correctness tests. The remaining test failures are generally about inference giving less precise types or allocations that didn't used to be there. I'm now gonna turn my attention to those. As a note the reviewers, this commit is not designed to be reviewed commit-by-commit at the moment. The intermediate commits have stupid bugs that I fixed in later commits, so probably best looked at as a whole, at least until I get around to do some squashing.

Keno commented:
Reduced test case:
```
julia> @noinline g(x, i) = x[i]
g (generic function with 1 method)

julia> f(x) = g(x, 2)
f (generic function with 1 method)

julia> code_typed(f, Tuple{Union{Nothing, Tuple{typeof(A), Int64}}})
1-element Array{Any,1}:
 CodeInfo(:(begin
      SSAValue(1) = (x isa Nothing)::Bool
      unless SSAValue(1) goto 5
      #temp# = $(Expr(:invoke, MethodInstance for g(::Nothing, ::Int64), :(Main.g), :(x), 2))::Union{Int64, Array{Int64,1}}
      goto 14
      5:
      SSAValue(2) = (x isa Tuple{Array{Int64,1},Int64})::Bool
      unless SSAValue(2) goto 10
      #temp# = $(Expr(:invoke, MethodInstance for g(::Tuple{Array{Int64,1},Int64}, ::Int64), :(Main.g), :(x), 2))::Union{Int64, Array{Int64,1}}
      goto 14
      10:
      goto 12
      12:
      (Base.error)("fatal error in type inference (type bound)")
      14:
      SSAValue(0) = #temp#
      return SSAValue(0)
  end)) => Union{Int64, Array{Int64,1}}
```
@vtjnash @JeffBezanson I was under the impression that the IPO stuff should have been able to figure this out. Is this a pass ordering problem wrt union splitting?

quinnj commented:
well, https://github.com/JuliaLang/julia/issues/23338 is still open?

Keno commented:
Certainly related, but I'm not sure it's the exact same issue. The one I just posted seems to be specifically needing the IPO stuff to work through union splitting, whereas the one you referenced seems to need another pass of inlining after the union-splitting.

vtjnash commented:
> Tuple{Union{Nothing, Tuple{Vector{Int64}, Int64}}

That's not ideal – we don't really want a type like this popping up anywhere, so we disable IPO while dealing with it

Keno commented:
Well that happens to be the return type from iterate so we have to deal
with it.

On Jan 12, 2018 8:45 PM, "Jameson Nash" <notifications@github.com> wrote:

> Tuple{Union{Nothing, Tuple{Vector{Int64}, Int64}}
>
> That's not ideal – we don't really want a type like this popping up
> anywhere, so we disable IPO while dealing with it
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/25261#issuecomment-357400231>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABO1l-klGKZXj9TBV6jPgprcyjNqo944ks5tKAq7gaJpZM4RLq9a>
> .
>

Keno commented:
(In case it wasn't clear the g in the example above is indexed_iterate)

Keno commented:
@vtjnash points out that the example I posted above can be fixed by type assert on the iterate call (though I do think inference should figure that out eventually). On to the next problem:
```
f() = (1, 1)
f(s) = i == 1 ? (1, 2) : nothing
function foo()
    next = f()
    while next !== nothing
        next = f(Core.getfield(next, 2))
    end
end
```
If you look at code_llvm, there's an allocation and a call to generic getfield in there. Hilariously if the call to `f()` can also return `nothing`, this works fine. What appears to be happening here is that, we look at the return value of `f()`, see it's `Tuple{Int, Int}`, then don't make use of the information that it's not nothing and then later discover that we actually need to widen the type of the slot to include the nothing, but don't revisit the check.

vtjnash commented:
This is an interesting one, since it stems from our current inference lattice design. In particular, we have that `Core.Inference.tmerge(Const(false), Conditional(Slot(n), T1, T2)) == Bool`, while ideally (in order to infer this) we would instead have preserved the knowledge that the previous type was actually `Conditional(Slot(n), T1, Union{})`. As long as we don't start inferring `isdefined(a)` as that type (which isn't likely to be very helpful anyways), we could define that `Conditional(Slot(n), T1, Union{})  ⊑ Const(true)` and teach inference to handle it accordingly.

Keno commented:
This is passing tests locally now, so

@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-357553411):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/86ba18a48ce057c169c2e76b2d4367120e2f21f9/279eb5e_vs_9924f79/report.md)
cc @ararslan

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Keno commented:
I think this one failed because of the iteration protocol itself. I'll put in the shim to make the old protocol work.

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-357559011):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/118a71aad0b8cb7ea5485307739e4499873bfdef/fd6c8a4_vs_60cd7cf/report.md)
cc @ararslan

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-357849804):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a46e80efe7809641b2216ef093354a1776de26ba/0e90e84_vs_5e2ff12/report.md)
cc @ararslan

ararslan commented:
Looks like the shim for supporting the current iteration protocol is causing ambiguity errors.

Keno commented:
Something is really odd about that. The error message indicates that it called the fallback start function while there's a perfectly good non-fallback start function that it could have called instead.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-358177676):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/22c35d7ccd44a3d6197d0d03d3f757e36784f7cc/eeb177a_vs_d548bf9/report.md)
cc @ararslan

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Keno commented:
I have rebased and squashed this. The major change from the previous iteration is that the `done` function I had in the original version of this PR is back as `isdone` and rather than throwing method errors now uses three-value logic. The idea behind this function is to check whether a mutable iterator is done without advancing its state, analogous to the `done` function in the old iteration protocol. However, unlike the previous `done` function, `isdone` is optional and uses 3VL. If the function returns `missing`, callers should fall back to just using `iterate(itr, state) === nothing`. To see why this is useful, consider `Stateful`'s use in string comparison:
```
function cmp(a::AbstractString, b::AbstractString)
    a === b && return 0
    a, b = Iterators.Stateful(a), Iterators.Stateful(b)
    for (c, d) in zip(a, b)
        c ≠ d && return ifelse(c < d, -1, 1)
    end
    isempty(a) && return ifelse(isempty(b), 0, -1)
    return 1
end
```
If `b` is shorter than `a`, we need to make sure not to consume any extra elements of `a` that we're not gonna surface to the user. `isempty` basically has the desired semantics, but if we were to use that in `zip`, we would be pessimzing some other iterators (because in general you need to perform an iteration to check isempty). `isdone` is basically a compromise to allow both. It allows mutable iterators with O(1) access to whether they're done or not to avoid spurious consumption, while simultaneously allowing iterators with expensive access to the next element to avoid answering the question. Note that 3VL has exactly the right semantics here, e.g. previously done on zip was defined as:
```
done(z::Zip, (sa, sb)) = done(z.a, sa) | done(z.b, sb)
```
which is the same definition as after this change (with `done`->`isdone`), but behaves correctly under 3VL (if either is `true`, you know you're done - if both are false you know you're not, otherwise you have to check). We also use `isdone` to implement a correct version of `isempty` for mutable iterators (in the absence of `isdone`, we'd have to thread `isempty` through iterator wrappers, do the implementation burden is the same).

I should point out though that @vtjnash dislikes the 3VL design for this function, because as he put it "3-VL logic is at least 3x harder". After thinking about this, I do still feel that 3VL is the correct way to model this, but I'm open to other suggestions.

There's one remaining todo item, which is to switch some remaining string code over to the new iteration protocol. There's also #26079, which is necessary for this to not have significant performance regressions in case of nested unions (e.g. iterating over `Vector{Union{Int, Nothing}}`). With respect to that the general thought was to merge this and tag a pre-alpha version of 0.7 even with known performance regressions (over current 0.7, not over 0.6), and then finish #26079, shortly thereafter for another alpha version, to give packages some more time with the new iteration interface (and the many other breaking changes).

Keno commented:
`isdone` is always safe to call in the 3VL design, so no trait is needed to guard it - if you think it'd be useful just call it. For immutable iterators, it is always safe to not implement. Iterators that are internally mutating can implement it to avoid being mutated by supported callers (currently zip and isempty). Most iterators are immutable so generally don't have to worry about it.

Keno commented:
Just to reiterate, use of `isdone` should be extremely rare - basically only to be used by callers if the next thing isn't gonna be getting the item (which is fairly rare - e.g. all iterator wrappers that only wrap a single iterator generally do not call `isdone`). It's also fairly rare for iterators to have to implement since it requires the combination of 1) reliably knowing whether you're done and 2) a side-effecting `iterate` method. If your `iterate` method doesn't have side effects, the compiler will generally just get rid of the unnecessary code, so there's no need to write an `isdone` method.

Overall, this PR reduces the number of required methods to be an iterator from 3 to 1 (two if you count the default method as a separate method), and leaves invariant (or reduces by one, depending on how you look at it), the number of methods you want to implement if you're writing a generic iterator wrapper (one that might be used on mutable iterators).

nalimilan commented:
It's interesting to see at least one area where 3VL is useful in Base itself!

vtjnash commented:
> To see why this is useful, consider Stateful's

I always thought that Stateful's implementation where it maintains a copy of the n+1st value and always advances the iterator too far was a design flaw with the old iteration protocol; and the point of changing to the new iteration protocol was to fix that. But that would render it impossible for `Stateful` to implement `isdone` given 3-VL. We could make a `PushBack` / `Lookahead` iterator that provides the additional caching behavior of `Stateful` however.

Keno commented:
I think your point about `Stateful` conflating two different things is very valid, though unfortunately, I don't think it gets us out of this mess. For example, suppose we split out `Stateful` and `LookAhead` and rewrote all the current uses of `Stateful` to `Stateful(LookAhead(a))` and used a reliable (as in always gives true/false if possible, errors otherwise) `isempty` in zip (which that would enable). In that case, I think you could either no longer write (or if it works, it wouldn't be efficient), things like the following:
```
for (i, x) in zip(1:10, filter(pred, itr))

end
```
which I don't think is an acceptable trade-off. One way to remidy that of course would be a `HasEfficientIsEmpty` trait, but it just feels like at that point you're back to 3VL, but externally implemented.

Keno commented:
Rebased and string resolution as described in #26133. One odd thing that happened is that sysimg time for SharedArrays jumped 10x, which I can't explain, so I need to look into that.

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

Keno commented:
@ararslan Kick it for me?

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-372077184) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b72380a98e22d5922227f94c079edcc1800ddc6e/17f30e6_vs_40f64e5/report.md). cc @ararslan

Keno commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

ararslan commented:
Nanosoldier hasn't been responding today. I'm going to do some maintenance on it and I'll trigger it here once I'm done.

ararslan commented:
Let's see how this goes.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-372147115) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d8195dcf4faf95c8e6000b5bf9a7800aff180252/beb73dc_vs_61bf935/report.md). cc @ararslan

Keno commented:
Ok, I looked at the most egregious remaining regressions here and they're essentially all due to problems with the old optimizer.

Keno commented:
Alright, those allocations are all tuples that the new optimizer should hack to pieces. For my future reference, the thing to benchmark for those failures is:
```
X = Y = sprandn(10, 12, 0.3)
Z = copy(X)
@allocated SparseArrays.HigherOrderFns._broadcast_notzeropres!(f, 1.0, Z, X, Y)
```

vtjnash commented on src/julia-syntax.scm:
```diff
@@ -1602,13 +1602,14 @@
                        (_while
-                        (call (top not_int) (call (core typeassert) (call (top done) ,coll ,state) (core Bool)))
-                        ,body))))))))
+                        (call (top not_int) (call (core ===) ,next (null)))
```
Interesting – why is `not_int` getting imported into `top`? This should be found in `(|.| (core Intrinsics) 'not_int)`, and not found in `top`.

JeffBezanson replied:
Base probably imports it.

vtjnash commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-382477789) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/495eabd14c974c0badb82b3a952c8e329bd3577a/beba563_vs_8e7034e/report.md). cc @ararslan

ararslan commented:
Wow, basically every benchmark is red for time, memory, or both. That's kind of impressive in its own way.

ViralBShah commented:
For such a large change, it seems that the worst case slowdown is 50%. That's quite encouraging actually. I was expecting it to be much worse.

Keno commented:
It is much worse, you probably read the results wrong. In any case it's without the new optimizer so not particularly surprising.

KristofferC commented:
`findprev` on Bool arrays got 3x faster. Progress! :D

Keno commented:
Planning to merge once CI passes.
One final go for good measure:

@nanosoldier `runbenchmarks(ALL, vs=":master")`

vtjnash commented:
I think ideally, we should deal with bug-fixes from merging SSAIR found by PkgDev (https://pkg.julialang.org/pulse.html) before we add a breaking change on top

Keno commented:
You'll have to convince people to delay tagging the alpha for that reason.

Keno commented:
Test failure is LLVM failing to vectorize because it doesn't like our loop new structure (early exit at the top). Shouldn't be too hard to fix. Will take a look in the morning. Alternatively, we may want to change the lowering after all to make the loop structure more obvious to LLVM.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-389708658) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e5506fb7c49ac2fa9ac9cbcdd0dcda38a9b6cae0/1f77d36_vs_d44944d/report.md). cc @ararslan

Keno committed:
Use loop inverted loop lowering

The primary idea of the new iteration protocol is that for
a function like:
```
function iterate(itr)
   done(itr) ? nothing : next(itr)
end
```
we can fuse the `done` comparison into the loop condition and
recover the same loop structure we had before (while retaining
the flexibility of not requiring the done function to be separate),
i.e. for
```
y = iterate(itr)
y === nothing && break
```
we want to have after inlining and early optimization:
```
done(itr) && break
y = next(itr)
```
LLVM performs this optimization in jump threading. However, we run
into a problem. At the top of the loop we have:
```
y = iterate
top:
%cond = y === nothing
br i1 %cond, %exit, %loop
....
```
We'd want to thread over the `top` block (this makes sense, since
by the discussion above, we need to merge our condition into the
loop exit condition). However, LLVM (quite sensibly) refuses to
thread over loop headers and since `top` is both a loop header
and a loop exit, we fail to perform the appropriate transformation.

However, there's a simple fix. Instead of emitting a foor loop as
```
y = iterate(itr)
while y !== nothing
    x, state = y
    ...
    y = iterate(itr, state)
end
```
we can emit it as
```
y = iterate(itr)
if y !== nothing
    while true
       x, state = y
       ...
       y = iterate(itr, state)
       y === nothing && break
    end
end
```
This transformation is known as `loop inversion` (or a special
case of `loop rotation`. In our case the primary benefit is
that we can fuse the condition contained in the initial `iterate`
call into the bypass if, which then lets LLVM understand our loop
structure.

Co-authored-by: Jeff Bezanson <jeff@juliacomputing.com>

Keno commented:
Fixing LLVM was too hard, I updated this to use the loop inverted lowering, which should fix the test case.

@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/25261#issuecomment-390026363) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/5a9bffd0ba871814bb5ab8999b0dd009b44389cf/bd41abd_vs_337ee84/report.md). cc @ararslan

Sacha0 commented:
![Boom!](https://media.giphy.com/media/mks5DcSGjhQ1a/giphy.gif)

JeffBezanson commented:
Do we, uh, have a handle on those 600x regressions?

Keno commented:
Yes, they trigger the deprecation warning for the old iteration protocol.

martinholters commented on base/essentials.jl:
```diff
@@ -783,3 +766,108 @@ ismissing(::Missing) = true
+    LegacyIterationCompat{T, typeof(val), typeof(state)}(val, state)
+end
+
+function next(itr::I, state::LegacyIterationCompat{I,T,S}) where {I,T,S}
```
This (and similar for `done`) are real ambiguity magnets for all iterators out there that define `next(it::MyIterator, state)`. Of course, `next` should never be called with arguments that hit the ambiguity, but `detect_ambiguities` is not amused. But I suppose there is no easy way to prevent that?
